### PR TITLE
[IMP] l10n_cn: add print option "Voucher"

### PR DESCRIPTION
**Pull request by yhu-odoo**
Add an additional report called "Voucher/记帐凭证" to print option in
account.move form view.
Also add new lib requirement "cn2an", we use this to convert numbers
to Chinese financial capital numbers.

Task 2179744

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by jco-odoo** [comment]
@julienlegros cn2an is in pip3 and I wonder if we can not just do the dependency somehow?  (in requirements.txt or just for l10n_cn)

File: addons/l10n_cn/models/__init__.py [hunk 1]
```diff
@@ -1,4 +1,3 @@
 # -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
```
Comment (jco-odoo) [inline addons/l10n_cn/models/__init__.py:2]: why?

File: addons/l10n_cn/models/account_move.py [hunk 2]
```diff
@@ -15,3 +17,18 @@ def _check_fapiao(self):
         for record in self:
             if record.fapiao and (len(record.fapiao) != 8 or not record.fapiao.isdecimal()):
                 raise ValidationError(_("Fapiao number is an 8-digit number. Please enter a correct one."))
+
+    @api.model
+    def convert_to_chinese_capital(self, number):
+        """Convert number to capital Chinese number for financial use."""
+        return an2cn(number, 'rmb')
+
+    def count_attachments(self):
```
Comment (jco-odoo) [inline addons/l10n_cn/models/account_move.py:26]: Why public?  Should not this be a private method?

File: addons/l10n_cn/models/account_move.py [hunk 3]
```diff
@@ -15,3 +17,18 @@ def _check_fapiao(self):
         for record in self:
             if record.fapiao and (len(record.fapiao) != 8 or not record.fapiao.isdecimal()):
                 raise ValidationError(_("Fapiao number is an 8-digit number. Please enter a correct one."))
+
+    @api.model
+    def convert_to_chinese_capital(self, number):
```
Comment (jco-odoo) [inline addons/l10n_cn/models/account_move.py:22]: why is this a public method?

File: addons/l10n_cn/views/report_voucher.xml [hunk 4]
```diff
@@ -0,0 +1,132 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
+        <template id="external_layout_boxed" inherit_id="web.external_layout_boxed" primary="True">
```
Comment (jco-odoo) [inline addons/l10n_cn/views/report_voucher.xml:4]: Is there a reason why we put a boxed layout? As normally, a report should work with any.

File: addons/l10n_cn/views/report_voucher.xml [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                    </div>
+                </div>
+                </div>
+            </xpath>
+        </template>
+
+        <template id="report_voucher_document">
+            <t t-set="o" t-value="o.with_context(lang=lang)" />
+            <t t-if="not company">
+                <!-- Multicompany -->
```
Comment (jco-odoo) [inline addons/l10n_cn/views/report_voucher.xml:19]: All this to find the company, seems really unconventional.  Where does it come from?  Normally you just take o.company_id (on the account_move (which is a related with the journal))

File: requirements.txt [hunk 6]
```diff
@@ -1,5 +1,6 @@
 Babel==2.6.0
 chardet==3.0.4
+cn2an==0.5.8
```
Comment (jco-odoo) [inline requirements.txt:3]: @d-fence Do we need anything else here in master?

**Review by jco-odoo — CHANGES_REQUESTED** [review]
@yhu-odoo Thanks a lot already for checking for the library.  It simplifies a lot.  It would be good if you can check the other comments still.

File: requirements.txt (hunk 6 above)
Comment (d-fence) [reply requirements.txt:3]: I'm sorry to say that I cannot accept this new requirement two reasons:

1. l10n_cn is an optional module and this requirements will be installed even if you don't install l10n_cn
2. AFAIK there is no Debian package for cn2an

Instead, you should try the import of cn2an and handle the case when the import fails
Look [here](https://github.com/odoo/odoo/blob/6abd50ae8e1b19339163c717705826544e222e13/addons/http_routing/geoipresolver.py#L5-L8) for example:

File: addons/l10n_cn/views/report_voucher.xml (hunk 4 above)
Comment (yhu-odoo) [reply addons/l10n_cn/views/report_voucher.xml:4]: @jco-odoo I did this according to the example provided by our Chinese partner. And according to my experience, I never see other layout used in China.

**Comment by yhu-odoo** [comment]
@jco-odoo Applied all the change, please take a look. Thanks.

File: addons/l10n_cn/static/src/js/action_manager_report.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    _executeReportAction: function (action, options) {
+        if (action.xml_id && action.xml_id == "l10n_cn.account_voucher_cn") {
+            this._rpc({
+                method: 'check_cn2an',
+                model: 'account.move',
+            }).then(res => {
+                if (!res) {
+                    var msg = _t("Unable to find cn2an on this system. The \"amount in words\" won't be shown.") + '<br><br><a href="https://pypi.org/project/cn2an/" target="_blank">cn2an · PyPI</a>';
+                    this.do_notify(_t('Voucher'), msg);
+                }
```
Comment (jco-odoo) [inline addons/l10n_cn/static/src/js/action_manager_report.js:22]: You could just have put nothing in the QWeb report and we describe this in the manifest somehow.  It is nice, but a bit overkill.  If we install it on the SaaS (I also think this is backportable to v14) they won't need to install manually anyways. @yhu-odoo

Comment (yhu-odoo) [reply addons/l10n_cn/static/src/js/action_manager_report.js:22]: Sorry I forget to remove this part. You could just remove this file and adapt the manifest. I just checked, it works fine.

**Comment by jco-odoo** [comment]
@yhu-odoo Is there a reason we could not do this PR in v14?

**Comment by yhu-odoo** [comment]
> @yhu-odoo Is there a reason we could not do this PR in v14?

**Pull request closed without merging**